### PR TITLE
fix: keep the fallback pipeline rendering while the window is hidden

### PR DESCRIPTION
**Pull request by CSantosM**
## Summary

In browsers without Insertable Streams (Firefox, Safari), a participant who **minimises or occludes their window** is seen frozen on their last frame by every remote participant, for as long as the window stays hidden. It only happens with a processor attached — the raw camera keeps flowing — and it recovers as soon as the window comes back.

The cause is the clock. The fallback path publishes `displayCanvas.captureStream()`, and the canvas is painted by a `requestAnimationFrame` loop. A hidden window is not painted, so it gets **no animation frames at all**, nobody repaints the canvas, and the captured track sits on its last frame. The modern `MediaStreamTrackProcessor` path is frame-driven and unaffected, which is why this is invisible on Chromium.

This replaces `requestAnimationFrame` with a worker timer in the fallback path only. Worker timers are exempt from the throttling browsers apply to hidden documents, and `canvas.captureStream()` keeps emitting for as long as something paints the canvas.

## Why the clock is the only thing that breaks

Measured in real Firefox 154 on Linux with the window minimised, using four loopback `RTCPeerConnection`s so the numbers are frames actually encoded and decoded, not just painted:

| while the window is minimised | rate |
| --- | --- |
| `requestAnimationFrame` callbacks | **0 / s** |
| worker `setTimeout` ticks | 31 / s |
| window `setInterval` ticks | ~1.9 / s (throttled) |
| source `<video>` element | keeps advancing |
| `canvas.captureStream()` (auto), while something paints | **29 fps** |

So capture and encode are perfectly healthy while hidden. Once the loop keeps ticking, the frozen-frame symptom disappears.

Two alternatives measured and rejected on Firefox 154, in case they come up in review:

- `captureStream(0)` + `videoTrack.requestFrame()` delivers **1 fps even while the window is visible**, so it is not usable as the capture mode.
- `canvas.transferControlToOffscreen()` makes `captureStream()` throw `NotSupportedError`, so moving the painting into a worker is not available either.

Firefox 153/154 exposes no `MediaStreamTrackGenerator`, `MediaStreamTrackProcessor` or `VideoTrackGenerator`, in window or worker scope, so `canvas.captureStream()` really is the only pixels-to-track path there, and the fallback path is the only path.

## Implementation notes

`createFrameTicker(intervalMs, onTick)` in the new `src/FrameTicker.ts` owns the clock, and `startRenderLoop` becomes a plain per-frame function.

**The worker is self-paced: one `setTimeout` per request, never a `setInterval`.** This matters more than it looks. The next tick is only armed once `onTick()` has returned, so when segmentation is slower than the interval (software WebGL, low-end hardware) the frame rate simply drops. My first attempt used `setInterval` in the worker and it wedged the page: ticks queued on the main thread faster than MediaPipe could consume them and the tab stopped responding to input. `requestAnimationFrame` never had that failure mode because it self-throttles, so taking over the clock means taking over the back-pressure too.

**The ticker runs at `2 × maxFps`.** The frame deadline in the loop is what caps the output rate; the ticker only samples the source, and sampling at twice `maxFps` is what keeps every deadline reachable on a 30 fps camera. `requestAnimationFrame` was already sampling at ~60 Hz against a default `maxFps` of 30, so this does not raise the sampling rate.

**The loop must never be able to die,** since the whole bug is a loop that stopped: the re-arm sits in a `finally`, so a throwing `onTick` cannot silently end the stream, and a `worker.onerror` falls back to `requestAnimationFrame` if the worker fails after construction. Environments that cannot spawn a blob worker at all (a CSP without `worker-src blob:`) also fall back to `requestAnimationFrame`, i.e. exactly today's behaviour, so the change cannot make any environment worse than it currently is.

The blob URL is revoked immediately after `new Worker(...)`. I verified the worker still loads and ticks with the URL already revoked on Chromium 151, Firefox 153 and WebKit 26.5, and a load failure would surface as the `onerror` fallback above rather than a dead loop.

The paused-source branch no longer re-arms the loop itself, since #128 has landed the rate-limited retry on `main` and the ticker keeps ticking on its own.

## Testing

Beyond the standalone probes above, I reproduced the bug and validated the fix in an end-to-end suite of a real application (two participants, real SFU): the publisher applies a background effect, and frames actually decoded by the viewer are counted with `getVideoPlaybackQuality().totalVideoFrames` before and while the publisher is hidden. Before this change the viewer went from 19 frames per 6 s window to 2; after it, the rate while hidden matches the visible baseline. A control run without a background effect is unaffected in both cases, which pins the regression to this path.

Note for anyone writing similar tests: Playwright cannot truly hide a window (`bringToFront()` leaves the other page `visible` in both Chromium and Firefox, headed or headless, and `Emulation.setPageVisibilityOverride` is gone from CDP), so the harness emulates the two observable effects — `document.visibilityState` plus `visibilitychange`, and neutered `requestAnimationFrame` — and deletes `MediaStreamTrackProcessor`/`MediaStreamTrackGenerator` before load so Chromium takes the fallback path.

## Known limitation, out of scope for this PR

On Firefox there is still a ~1.2 s freeze at the **moment** of minimising, which this change cannot address: Firefox throttles the document's refresh driver to ~1 Hz *before* `visibilitychange` fires, and `canvas.captureStream()` rides on that clock. Measured in 200 ms buckets across minimise/restore cycles, the render loop never stalls (0 ms) and the raw camera never gaps (0 ms), yet the captured track delivers ~1 fps for ~1.2 s while the document still reports itself `visible`; it returns to full rate as soon as the document is marked hidden. Restoring the window costs only ~400 ms of reduced rate. So this fixes "frozen for as long as the window is hidden", not the transition itself, which looks like a Firefox issue rather than something this library can work around.

### CODE REVIEW

**Comment by CSantosM** [comment]
Ping @1egoman

File: src/FrameTicker.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * exempt from that throttling, and `canvas.captureStream()` keeps emitting for as long as
+ * something paints the canvas.
+ *
+ * Where a blob worker is not available (for example a Content-Security-Policy without
+ * `worker-src blob:`), falls back to `requestAnimationFrame`, which stalls while hidden.
+ */
+export function createFrameTicker(intervalMs: number, onTick: () => void): FrameTicker {
+  let worker: Worker;
+
+  try {
```
Comment (1egoman) [inline src/FrameTicker.ts]: Just generally, I'm a bit confused why you have both `createFrameTicker` and `createAnimationFrameTicker`. I would think given the throttling behavior of `requestAnimationFrame` it would make more sense to move to an approach with a non thottling timer like `createFrameTicker` everywhere?

File: src/FrameTicker.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    const workerUrl = URL.createObjectURL(
+      new Blob([TICKER_WORKER_SOURCE], { type: 'text/javascript' }),
+    );
```
Comment (1egoman) [inline src/FrameTicker.ts:34]: Note that this pattern is something we've found to be unreliable in the js sdk (due to variances in build systems) and have purposefully avoided it. If you do need a webworker, target something [closer to how the existing e2ee worker functions in the js sdk](https://docs.livekit.io/frontends/build/sessions/#end-to-end-encryption). This is fairly cumbersome though and would result in a breaking api interface change which I don't love.

This generally makes me wonder if there's alternate way to do this without a worker. I found [this on stack overflow](https://stackoverflow.com/a/30572290), maybe it could be a viable alternate solution to get a non throtling timer source?

File: src/ProcessorWrapper.ts [hunk 3]
```diff
@@ -351,10 +356,11 @@ export default class ProcessorWrapper<
           this.log.error('Error in render loop:', e);
         }
       }
-      this.animationFrameId = requestAnimationFrame(renderLoop);
     };
 
-    this.animationFrameId = requestAnimationFrame(renderLoop);
+    // Ticking at twice maxFps leaves the frame gate above room to hit its interval: pacing the
+    // ticker at exactly maxFps makes normal jitter miss it every other tick and halves the rate.
+    this.frameTicker = createFrameTicker(1000 / (this.maxFps * 2), renderFrame);
```
Comment (1egoman) [inline src/ProcessorWrapper.ts]: I would think this would in effect make the frame rate run twice as fast in normal usage, which makes `maxFps` no longer really mean what it says. Maybe a better approach here would be to increase the default max frame rate option from `30` to `45` or higher? I hesitate to think double would be needed, and since a user can override it directly they can set it to whatever they would need.

File: src/ProcessorWrapper.ts [hunk 4]
```diff
@@ -274,10 +276,13 @@ export default class ProcessorWrapper<
...
-        this.sourceDummy.play().then(() => {
-          this.animationFrameId = requestAnimationFrame(renderLoop);
-        });
+        if (performance.now() - lastResumeAttempt > 1000) {
+          lastResumeAttempt = performance.now();
+          this.log.warn('Video is paused, trying to play');
+          this.sourceDummy
+            .play()
+            .catch((e) => this.log.warn('Could not resume the source video:', e));
+        }
```
Comment (1egoman) [inline src/ProcessorWrapper.ts:297]: I am a bit confused here, the old code had `this.sourceDummy.play().then(() => { this.animationFrameId = requestAnimationFrame(renderLoop); });`.

The new code doesn't seem to ever kick off `frameTicker` directly in this path. Is the idea that this is no longer needed because the `frameTicker` is just always running in the background?

I bring this up mostly because I am worried that this was important to have here for some sort of edge case (like maybe the timer needs to be restarted after the video is unpaused / etc).

**Review by 1egoman — COMMENTED** [review]
Generally the problem makes sense and I agree it's not something this plugin handles well today, but I am not fully sure I think this is the right solution

File: src/FrameTicker.ts (hunk 1 above)
Comment (CSantosM) [reply src/FrameTicker.ts]: Fair, the naming was doing me no favours. They are not two options: the worker is the clock, and `requestAnimationFrame` is only there for environments where a blob worker cannot run at all (a CSP without `worker-src blob:`), where the alternative is no loop rather than a throttled one. Renamed it to `createRequestAnimationFrameTicker` and moved it below the exported one so it reads that way.

It does earn its keep, though: I checked, and a CSP that blocks the worker does not throw, it fires `onerror` asynchronously (same in chromium, firefox and webkit), so the fallback is the only thing
catching that case. I added a start timeout too, in case some engine does neither.

File: src/ProcessorWrapper.ts (hunk 3 above)
Comment (CSantosM) [reply src/ProcessorWrapper.ts]: It does not double the output: the ticker samples, the gate caps. With `maxFps: 30` on a 30 fps camera:

| clock | firefox | chromium |
| --- | --- | --- |
| `requestAnimationFrame` (today) | 22.4 fps | 20.8 fps |
| worker at 2x | 28.1 fps | 26.3 fps |

Never above 30. rAF already samples at ~60 Hz against a default `maxFps` of 30, so this lowers the sampling rate rather than raising it.

You were right that the multiplier should not matter, though, and chasing that found the real bug:
`timeSinceLastFrame >= minFrameInterval` quantises the output against any periodic clock, which is why `maxFps: 30` measures 21 to 23 fps on `main` today. I swapped it for a deadline and it is now 30.0 fps with either clock, pushed in 8050931.

File: src/ProcessorWrapper.ts (hunk 4 above)
Comment (CSantosM) [reply src/ProcessorWrapper.ts:297]: Right, that is the idea: the ticker is independent of the render function now, so the early return just skips a frame and the next tick arrives anyway. The old re-arm had to be there because `requestAnimationFrame` is one-shot and the early return skipped the one at the end of the loop.

Correction to what I first wrote here: #128 landed the same rate-limited retry on `main` while this PR was open, and I have merged it. So this hunk is now only dropping the `requestAnimationFrame(renderLoop)` re-arm, because with the ticker owning the clock, re-arming here would start a second loop on top of it.

File: src/FrameTicker.ts (hunk 2 above)
Comment (CSantosM) [reply src/FrameTicker.ts:34]: I think the SDK's pain does not reach this one. There the worker is a real module the bundler has to resolve and emit as a chunk (`new Worker(new URL(...))`), which is exactly what build systems disagree about. Here it is a one-line string constant, nothing to resolve, nothing to emit. What can actually block it is a CSP, and that falls back to rAF at full tick rate in all three engines.

I did try the stackoverflow one. It keeps ticking while hidden, so it works, but the rate is `sampleRate / bufferSize`, so it depends on the user's sound card (23.3 Hz at 48 kHz, 21.3 at 44.1 kHz) and cannot be paced at `maxFps`. And the render callback ends up running inside the audio callback: with an 80 ms frame it fell to 12.4/s with the device underrunning on every frame.
`ScriptProcessorNode` is also deprecated, and AudioWorklet puts the separate-file problem right back.

If the blob URL is still a no, I am happy to add an optional `workerFactory` so an app can pass its own worker, default unchanged. Just say the word.

**Comment by CSantosM** [comment]
Thanks for looking at this. Pushed a follow-up with the three things the review turned up: the ticker rename, a start watchdog on the worker (a CSP-blocked worker reports itself through `onerror` rather than by throwing), and the frame gate is now a deadline, which is where your `maxFps` comment led: the old comparison was costing 7 to 9 fps on `main` today.

On the premise, I redid the measurement on a real Firefox window minimised by the window manager, counting frames encoded over a loopback peer connection:

| canvas painted from | encoded fps while minimised | while visible |
| --- | --- | --- |
| `requestAnimationFrame` | 1.0 | 18 to 22 |
| worker ticker | 21.4 | 21 to 22 |

Worth spelling out the mechanism, since it rules out the obvious alternative: what collapses is the refresh driver, which drops to about 1 Hz for a window that is not being painted, and on Firefox that happens before `visibilitychange` fires. So a fix keyed on the visibility event would still leave a gap at the transition.

File: src/FrameTicker.ts (hunk 2 above)
Comment (1egoman) [reply src/FrameTicker.ts:36]: Thought about this overnight - I can't really think of another way to solve this problem and given the CSP fallback, the fallout is less severe than other webworkers we have elsewhere in the stack. It is fairly well encapsulated as well which at least limits complexity.

One additional thing - I put a LLM on this and had it run a bunch of tests in chrome, safari, and firefox to test this solution and a few others. It had this suggestion to replace `requestAnimationFrame` in `createAnimationFrameTicker` with `setTimeout`, which I think is probably a good idea:
> Replace requestAnimationFrame with setTimeout, unconditionally. One line, one path, no worker, no AudioContext. For this loop setTimeout is a strict superset of rAF: identical 30.0 fps visible, and 1 fps instead of 0 when hidden. That turns "frozen until you come back" into "slow but live", which for a minimised publisher is most of the user-visible fix.

I think this would at least make the strict CSP case slightly better? Any concerns with this approach?

File: src/FrameTicker.ts (hunk 1 above)
Comment (1egoman) [reply src/FrameTicker.ts]: Ah, I think I missed that one was exported and one was not. Ok, I think this makes sense.

**Comment by 1egoman** [comment]
Leaving this mostly as a comment for others at LiveKit: Reviewing this has made me realize we really desperately need e2e tests in this repo. It's really hard to get everything spun up across many devices (different browsers, mobile devices, etc) to test all these edge cases now. I think this PR and #128 have really made this abundantly clear and I'm going to start working on something in the background.

File: src/FrameTicker.ts (hunk 2 above)
Comment (CSantosM) [reply src/FrameTicker.ts:36]: Great!
Yes, that sounds fine to me :)

**Comment by CSantosM** [comment]
Pushed 984e1af, which changes the fallback path: where the blob worker cannot run it now uses a self-paced main thread timer instead of `requestAnimationFrame`. On reflection rAF made a poor degraded path, since it is the clock this PR exists to get away from.

Measured under `worker-src 'none'`, the timer holds the same rate as rAF did while the window is visible (61.0 ticks/s on chromium and 60.6 on firefox, against 60.0 and 60.3 for rAF), and a hidden document throttles a window timer to about 1 Hz rather than stopping it. So where a CSP blocks the worker, the published track now goes slow instead of freezing. It re-arms in a `finally` after each frame, same as the worker, so a slow frame lowers the rate instead of queueing work.

Side effect for your first comment: the two clocks in the file are now the same shape, one on a worker and one on the main thread, which should read better than the previous pairing.

**Comment by 1egoman** [comment]
@CSantosM In getting ready to merge this, I had a LLM go through this and it pointed out these two issues. Would you be able to take a look at these items? The first one sounds like potentially a non trivial issue with a fairly straightforward fix, and the second one sounds like post `requestAnimationFrame` -> `setTimeout` migration, explicitly validating `maxFps > 0` and maybe throwing an error at initialization time would be worthwhile.

Other than that these below items, I think this PR looks good to me and I should be able to merge within the next few days!

### 1. The start watchdog can kill a healthy worker, silently reinstating the bug this PR fixes. src/FrameTicker.ts:70

ticked only flips inside worker.onmessage, which needs a free main thread. The watchdog is a window timer at Math.max(1000, 16.7 * 10) = 1000ms. If the main thread is blocked for ≥1s any time after createFrameTicker returns — MediaPipe warm-up, the app's own re-render after setProcessor resolves, a slow device — both the timer task and the pending worker message become runnable at once and the UA may run the timer first. I reproduced this in Safari 26 with a 1.5s block: FALLBACK ENGAGED: worker did not start, on a worker that was fine. Because useTimeoutFallback starts with if (fallback) return, there is no recovery path — the processor spends the rest of its life on the window timer, which is ~1 fps while hidden, announced only by a log.warn.

Worth noting how narrow the watchdog's real job is: I served the page under Content-Security-Policy: worker-src 'none' and in both engines construction does not throw and onerror fires at 13ms (Safari) / 19ms (Firefox). So the author's premise is right, but the error always arrives roughly 60× sooner than the watchdog — the watchdog only guards a hypothetical engine that does neither, and in exchange it can misfire on a real one.

The cleanest fix also removes a branch: start the timeout ticker immediately, and stop it on the worker's first tick. No threshold to tune, no gap at startup if the worker really is blocked, no ticked/startTimeout at all. Brief double-sampling is harmless because the frame deadline caps output at maxFps — that's the same property the 2× sampling already relies on.

### 2. maxFps: 0 wedges the page. src/ProcessorWrapper.ts:375

maxFps comes straight from the public options and is never validated (options.maxFps ?? 30). With 0 — a plausible spelling of "unlimited" — the interval is Infinity, which WebIDL coerces to 0 for setTimeout's long delay (I confirmed the coercion), so the worker posts back with no delay and hammers the main thread. Meanwhile minFrameInterval is Infinity, so nextFrameDue becomes Infinity after the first frame and the track freezes — pegged CPU, frozen output. Negative values behave the same. Before this PR the same config just idled. Math.max(1, maxFps) in the constructor covers it.

**Comment by CSantosM** [comment]
Thanks @1egoman , both are real. Fixed in 37074fd and 6400944

### 1. Start watchdog

Confirmed. A worker's startup runs through the main thread: with the main thread blocked for 800 ms right after `createFrameTicker` returned, the first tick arrived at ~820 ms instead of ~25 ms, in all three engines. So any block at least as long as the watchdog, right after the ticker is created, always runs the overdue watchdog first.

I went with the design you describe: the window timer runs from the start, the worker's first message stops it, and `worker.onerror` recreates it if the worker fails later. `ticked`, `startTimeout` and the `if (fallback) return` guard are gone. 

### 2. `maxFps: 0`

Confirmed: `setTimeout(fn, Infinity)` fires at 0 ms in all three engines, the worker round-trips 8k to 12k times per 500 ms (29 at `maxFps: 30`), and the output freezes after the first frame. I took the constructor-time validation you suggested rather than a clamp: `RangeError` unless `maxFps` is a finite number greater than 0, which also covers `NaN` and `Infinity` (`Math.max(1, x)` lets both through to the same tight loop). Documented in the option's TSDoc and in the changeset.

**Review by 1egoman — APPROVED** [review]

**Comment by 1egoman** [comment]
Thanks for addressing all my notes on this @CSantosM, just merged!
